src/plugins/data_views/common/data_views/data_views.ts

```diff
@@ -999,11 +1002,16 @@ export class DataViewsService {
         return cachedDataView;
       }
 
-      return this.dataViewCache.set(spec.id, doCreate());
+      const dataViewPromise = doCreate();
+
+      this.dataViewCache.set(spec.id, dataViewPromise);
+
+      return dataViewPromise;
     }
 
     const dataView = await doCreate();
-    return this.dataViewCache.set(dataView.id!, Promise.resolve(dataView));
+    this.dataViewCache.set(dataView.id!, Promise.resolve(dataView));
+    return dataView;
```

Comment:
I can see that it was similar before but can't help but ask why 1005-1009 and 1012-1014 lines have different code style and return type :)

Reply:
The return type of `set` differs between new and old code. Previously it returned the value being set but the Map object returns a reference to the Map object.

Is that what you're asking about? I can't help but wonder if I'm overlooking something.

Reply:
I mean why we use `await` inside `if` statement block and no `await` (1005-1009) but `Promise.resolve(dataView)` outside it (1012-1014). Also we return `dataViewPromise` from the `if` statement block but `dataView` outside it.

Reply:
We need to have an id to add the data view promise to the cache. When we have the id, we add the `Promise<DataView>` to the cache before its resolved. However, when we don't have an id, we need to wait until its been created to get its id. While we're returning `A | Promise<A>` from inside the code, the return type of the function remains `Promise<A>`